### PR TITLE
Remove binary curves

### PR DESCRIPTION
**Pull request by wbl**
Not performant without dedicated hardware that never emerged, these just caused CVEs.

## Checklist
Thank you for your contribution. Please answer the following:

- [x] I acknowledge that I am authorized to submit this code under
the terms of the [Apache License](https://www.apache.org/licenses/LICENSE-2.0)
- [x] All new public APIs are documented
- [x] All new public functions have tests
- [x] All new public functions have fuzzing tests

### CODE REVIEW

**Comment by wbl** [comment]
The future of libcrypto.num is uncertain, and I'm not sure what workaround we have to preserve limited symbol visiblity without this pain, or what OS requires it.

**Comment by wbl** [comment]
Just realized we still have docs to clean up from this, which i shall do tomorrow

**Comment by botovq** [comment]
Apart from docs and various infra files, there are `crypto/bn/asm/*-gf2m.pl` and various prototypes in `include/crypto/ec_local.h`. A grep for `GF2m` should come up empty. There are some GF2m-only method members, like `field_div()` in `EC_METHOD`. Switching to C99 named initializers helps finding and getting rid of those. `EC2M_GETTABLE_DOM_PARAMS` (now defined to nothing) can probably go.

Also, `OPENSSL_NO_EC2M` should be defined somewhere in a public header so that the guarded code gets disabled in consumers.

**Comment by richsalz** [comment]
Looks good so far.  @botovq raises some good points (no need to do the initializers right now tho). Approved this if you promise to do the other stuff :)

**Comment by wbl** [comment]
Yeah, there definitely more that needs doing and I had not thought about exposing OPENSSL_NO_EC2M to applications. That's a very good idea. Might have been a bit early to open, but shouldn't be too long before it is really ready for review.

**Comment by wbl** [comment]
I think I'll deal with docs and cleaning up the rest of the functions and definitions and defines, but leave named initializers for a later PR. Maybe some field cleanup here?

**Comment by richsalz** [comment]
That plan – save initializers for later – makes complete sense.

**Comment by wbl** [comment]
I think its ready for review now.

File: crypto/ec/ec_asn1.c
Comment (botovq) [inline crypto/ec/ec_asn1.c:224]: I think the opening brace belongs on the previous line

File: crypto/ec/ec_asn1.c
Comment (botovq) [inline crypto/ec/ec_asn1.c:493]: same here

File: crypto/ec/ec_asn1.c
Comment (botovq) [inline crypto/ec/ec_asn1.c:497]: Join with previous closing brace

File: crypto/ec/ec_oct.c
Comment (botovq) [inline crypto/ec/ec_oct.c:41]: another one of these

File: crypto/ec/ec_oct.c
Comment (botovq) [inline crypto/ec/ec_oct.c:77]: and here

File: crypto/ec/ec_oct.c
Comment (botovq) [inline crypto/ec/ec_oct.c:102]: again

File: doc/man3/EC_GROUP_new.pod
Comment (botovq) [inline doc/man3/EC_GROUP_new.pod:119]: i think the docs usually call these 'aliases' (there are more, which i won't flag)

File: include/openssl/configuration.h.in
Comment (botovq) [inline include/openssl/configuration.h.in:76]: Makes sense here. Should it be wrapped into a `#ifndef` guard, so people who set it don't redefine it?

File: providers/fips/self_test_data.inc
Comment (botovq) [inline providers/fips/self_test_data.inc:1451]: The rest of the file places the `#endif ` after the closing brace

File: test/recipes/15-test_genec.t
Comment (botovq) [inline test/recipes/15-test_genec.t:83]: weird placement for the closing bracket.

File: doc/man3/EC_GROUP_new.pod
Comment (botovq) [inline doc/man3/EC_GROUP_new.pod:72]: Still more GF2m in this file (also in EC_GFp_simple_method.pod and EC_POINT_new.pod that should be mopped up.

**Review by botovq — COMMENTED** [review]
Generally looks good. Some nits seen when scanning through. Not sure how deeply you care about things like `util/`

**Comment by wbl** [comment]
Looks also like some doc nits need fixing. Thanks @botovq for the careful review: I will find time today/tomorrow to clean up those nits and land this.

**Comment by wbl** [comment]
First I will fix up the check_docs, then correct all the nits and other things that have been identified.

**Comment by wbl** [comment]
Now nitting time.

File: include/openssl/configuration.h.in
Comment (wbl) [reply include/openssl/configuration.h.in:76]: I'm not quite sure what difference would happen there, but you probably have more experience with that kind of mess than I do. Will add.

**Comment by botovq** [comment]
Oh, it's just to avoid redefining a define that is already set for some
reason that shouldn't really happen. I don't think guards can hurt in
any way. And if you look at the generated configuration.h in an openssl,
all the defines are guarded.

**Comment by wbl** [comment]
Nits cleaned up from the conversation: will look back to be sure that's all of them.

**Review by botovq — APPROVED** [review]
LGTM now. Some GF2m stuff remaining in util/missing* and gf2m in providers/fips* but I'm not sure you want to touch either of these right now.

**Comment by wbl** [comment]
Yeah, I don't personally understand what util/missing is generated by.

File: doc/man3/EC_GROUP_new.pod
Comment (botovq) [inline doc/man3/EC_GROUP_new.pod:180]: this 2 should go away

**Comment by botovq** [comment]
> Yeah, I don't personally understand what util/missing is generated by.

I wonder if it's not manually maintained. The doc-nits run checks if all symbols are documented except those listed in this file. So you're only allowed to remove from this list, but not to add.

If you run `make doc-nits` after a build, it shows this,
```shell
$ make doc-nits
perl ./util/find-doc-nits -c -n -l -e
*** WARNING: empty section in previous paragraph at line 186 in file doc/man3/EC_GROUP_new.pod
```
so it spots the spurious 2 you just fixed.

If I fix this and delete the first line of `util/missingcrypto.txt`, it whines about the undocumented `ACCESS_DESCRIPTION_it`:

```shell
$ make doc-nits
perl ./util/find-doc-nits -c -n -l -e
*** WARNING: empty section in previous paragraph at line 185 in file doc/man3/EC_GROUP_new.pod
crypto: function ACCESS_DESCRIPTION_it(3) undocumented
# 1 libcrypto names are not documented
*** Error 1 in /tmp/quictls (Makefile:4074 'doc-nits')
```

**Comment by wbl** [comment]
But we chopped them out of the code though? Maybe it uses the old ones as source? Anyway that mystery can wait

**Review by botovq — APPROVED** [review]

**Comment by wbl** [comment]
Merge button still greyed out which I don't get given your approval

**Comment by botovq** [comment]
I suppose you need a project member's approval, which I'm not.

**Comment by wbl** [comment]
@botovq invited you with I think the right minimal access amount.

**Review by botovq — APPROVED** [review]

**Comment by botovq** [comment]
Accepted and approved again.

**Comment by richsalz** [comment]
Is there some switch we need to flip so that @botovq is a first-class approver?  We need to do that.

**Comment by richsalz** [comment]
@botovq is right: util/missing is used to suppress warnings about undocumented functions and you remove things (I forget if find-doc-nits checks if something in missing is documented) by hand.

**Comment by botovq** [comment]
@richsalz Thanks. I think @wbl gave methe right permissions for now. I also had a clickable merge button if I'm not mistaken. We can revisit should the need arise.

**Comment by wbl** [comment]
You know what I forgot to do? Remove the option from the configure script.

**Comment by botovq** [comment]
Ha! ec2m. Didn't look for that one.

**Pull request merged**
